src/Storages/ObjectStorage/StorageObjectStorageStableTaskDistributor.cpp

```diff
@@ -0,0 +1,175 @@
+#include "StorageObjectStorageStableTaskDistributor.h"
+#include <Common/SipHash.h>
+#include <consistent_hashing.h>
+#include <optional>
+
+namespace DB
+{
+
+StorageObjectStorageStableTaskDistributor::StorageObjectStorageStableTaskDistributor(
+    std::shared_ptr<IObjectIterator> iterator_,
+    std::optional<std::vector<std::string>> ids_of_nodes_)
+    : iterator(std::move(iterator_))
+    , connection_to_files(ids_of_nodes_.has_value() ? ids_of_nodes_.value().size() : 1)
+    , ids_of_nodes(ids_of_nodes_)
+    , iterator_exhausted(false)
+{
+}
+
+std::optional<String> StorageObjectStorageStableTaskDistributor::getNextTask(size_t number_of_current_replica)
+{
+    LOG_TRACE(
+        log,
+        "Received a new connection from replica {} looking for a file",
+        number_of_current_replica
+    );
+
+    // 1. Check pre-queued files first
+    if (auto file = getPreQueuedFile(number_of_current_replica))
+        return file;
+
+    // 2. Try to find a matching file from the iterator
+    if (auto file = getMatchingFileFromIterator(number_of_current_replica))
+        return file;
+
+    // 3. Process unprocessed files if iterator is exhausted
+    return getAnyUnprocessedFile(number_of_current_replica);
+}
+
+size_t StorageObjectStorageStableTaskDistributor::getReplicaForFile(const String & file_path)
+{
+    if (!ids_of_nodes.has_value())
+        return 0;
+
+    /// Trivial case
+    if (ids_of_nodes.value().size() < 2)
+        return 0;
+
+    /// Rendezvous hashing
+    size_t best_id = 0;
+    UInt64 best_weight = sipHash64(ids_of_nodes.value()[0] + file_path);
+    for (size_t id = ids_of_nodes.value().size() - 1; id > 0; --id)
```

Comment:
Why reverse?

Reply:
Change to direct order to avoid confusing.